File: yep/src/main/java/catchla/yep/loader/ConversationsLoader.java
package catchla.yep.loader;

import android.accounts.Account;
import android.content.Context;

import org.mariotaku.sqliteqb.library.Expression;
import org.mariotaku.sqliteqb.library.OrderBy;

import catchla.yep.model.Conversation;
import catchla.yep.model.Message;
import catchla.yep.provider.YepDataStore.Conversations;
import catchla.yep.util.Utils;

/**
 * Created by mariotaku on 14-8-5.
 */
public class ConversationsLoader extends ObjectCursorLoader<Conversation> {

    public ConversationsLoader(Context context, Account account, boolean loadCirclesOnly) {
        super(context, Conversation.Indices.class, Conversations.CONTENT_URI, Conversations.COLUMNS,
                null, null, new OrderBy(Conversations.UPDATED_AT, false).getSQL());
        final Expression where;
        final String[] whereArgs;
        if (loadCirclesOnly) {
            where = Expression.and(Expression.equalsArgs(Conversations.ACCOUNT_ID),
                    Expression.equalsArgs(Conversations.RECIPIENT_TYPE));
            whereArgs = new String[]{Utils.getAccountId(context, account),
                    Message.RecipientType.CIRCLE};
        } else {
            where = Expression.and(Expression.equalsArgs(Conversations.ACCOUNT_ID));
            whereArgs = new String[]{Utils.getAccountId(context, account)};
        }
        setSelection(where.getSQL());
        setSelectionArgs(whereArgs);
    }

}

File: src/main/kotlin/catchla/yep/loader/ConversationsLoader.kt
package catchla.yep.loader

import android.accounts.Account
import android.annotation.SuppressLint
import android.content.Context
import android.database.Cursor
import android.database.MergeCursor
import catchla.yep.model.Conversation
import catchla.yep.model.ConversationCursorIndices
import catchla.yep.model.ConversationTableInfo
import catchla.yep.model.Message
import catchla.yep.provider.YepDataStore.Conversations
import catchla.yep.util.Utils
import org.mariotaku.sqliteqb.library.Expression
import org.mariotaku.sqliteqb.library.OrderBy
import java.util.*

/**
 * Created by mariotaku on 14-8-5.
 */
class ConversationsLoader(context: Context, account: Account, recipientType: String?, private val displayCircleEntry: Boolean) : ObjectCursorLoader<Conversation>(context, ConversationCursorIndices::class.java, Conversations.CONTENT_URI, ConversationTableInfo.COLUMNS, null, null, OrderBy(Conversations.UPDATED_AT, false).sql) {

    private val accountId: String

    init {
        accountId = Utils.getAccountId(context, account)
        val whereConditions = ArrayList<Expression>()
        val whereArgs = ArrayList<String>()
        whereConditions.add(Expression.equalsArgs(Conversations.ACCOUNT_ID))
        whereConditions.add(Expression.notEqualsArgs(Conversations.CONVERSATION_ID))
        whereArgs.add(accountId)
        whereArgs.add(Conversation.generateId(Message.RecipientType.USER, accountId))
        if (recipientType != null) {
            whereConditions.add(Expression.equalsArgs(Conversations.RECIPIENT_TYPE))
            whereArgs.add(recipientType)
        }
        selection = Expression.and(*whereConditions.toTypedArray()).sql
        selectionArgs = whereArgs.toTypedArray()
    }

    override fun query(): Cursor {
        if (!displayCircleEntry) return super.query()
        val original = super.query()
        val where = Expression.and(Expression.equalsArgs(Conversations.ACCOUNT_ID),
                Expression.equalsArgs(Conversations.RECIPIENT_TYPE))
        val whereArgs = arrayOf(accountId, Message.RecipientType.CIRCLE)
        @SuppressLint("Recycle") val chatEntry = context.contentResolver.query(mUri, mProjection,
                where.sql, whereArgs, mSortOrder)
        return MergeCursor(arrayOf(chatEntry, original))
    }
}
